cmd/test_load.go

```diff
@@ -192,59 +171,103 @@ func detectTestType(data []byte) string {
 	consolidatedConfig, err := getConsolidatedConfig(gs, cliConfig, lt.initRunner.GetOptions())
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	gs.logger.Debug("Parsing thresholds and validating config...")
 	// Parse the thresholds, only if the --no-threshold flag is not set.
 	// If parsing the threshold expressions failed, consider it as an
 	// invalid configuration error.
-	if !lt.runtimeOptions.NoThresholds.Bool {
+	if !lt.preInitState.RuntimeOptions.NoThresholds.Bool {
 		for metricName, thresholdsDefinition := range consolidatedConfig.Options.Thresholds {
 			err = thresholdsDefinition.Parse()
 			if err != nil {
-				return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+				return nil, errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
 			}
 
-			err = thresholdsDefinition.Validate(metricName, lt.metricsRegistry)
+			err = thresholdsDefinition.Validate(metricName, lt.preInitState.Registry)
 			if err != nil {
-				return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+				return nil, errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
 			}
 		}
 	}
 
 	derivedConfig, err := deriveAndValidateConfig(consolidatedConfig, lt.initRunner.IsExecutable, gs.logger)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	return &loadedAndConfiguredTest{
+		loadedTest:         lt,
+		consolidatedConfig: consolidatedConfig,
+		derivedConfig:      derivedConfig,
+	}, nil
+}
+
+// loadedAndConfiguredTest contains the whole loadedTest, as well as the
+// consolidated test config and the full test run state.
+type loadedAndConfiguredTest struct {
+	*loadedTest
+	consolidatedConfig Config
+	derivedConfig      Config
+}
+
+func loadAndConfigureTest(
+	gs *globalState, cmd *cobra.Command, args []string,
+	cliConfigGetter func(flags *pflag.FlagSet) (Config, error),
+) (*loadedAndConfiguredTest, error) {
+	test, err := loadTest(gs, cmd, args)
+	if err != nil {
+		return nil, err
 	}
 
-	lt.consolidatedConfig = consolidatedConfig
-	lt.derivedConfig = derivedConfig
+	return test.consolidateDeriveAndValidateConfig(gs, cmd, cliConfigGetter)
+}
+
+func (lct *loadedAndConfiguredTest) buildTestRunState(
+	configToReinject lib.Options,
+) (*lib.TestRunState, error) {
+	// This might be the full derived or just the consodlidated options
+	if err := lct.initRunner.SetOptions(configToReinject); err != nil {
+		return nil, err
+	}
+
+	// TODO: init atlas root node, etc.
 
-	return nil
+	return &lib.TestRunState{
+		TestPreInitState: lct.preInitState,
+		Runner:           lct.initRunner,
+		Options:          lct.derivedConfig.Options, // we will always run with the derived options
+	}, nil
 }
 
-type syncWriter struct {
-	w io.Writer
+type syncWriteCloser struct {
+	w io.WriteCloser
 	m sync.Mutex
 }
 
-func (cw *syncWriter) Write(b []byte) (int, error) {
+func (cw *syncWriteCloser) Write(b []byte) (int, error) {
 	cw.m.Lock()
 	defer cw.m.Unlock()
 	return cw.w.Write(b)
 }
+
+func (cw *syncWriteCloser) Close() error {
```

Comment:
nit: The idea behind this not having a `Close` method was that it makes no sense for something in the tests to be able to just close it.

Reply:
ah, makes perfect sense, will restore the old behavior

Reply:
Resolved by https://github.com/grafana/k6/pull/2627/commits/a75766b398ef3ed87628362e22f715a50d70a525